packages/fuels-core/src/code_gen/resolved_type.rs

```diff
@@ -0,0 +1,115 @@
+use crate::code_gen::custom_types_gen::extract_custom_type_name_from_abi_property;
+use fuels_types::errors::Error;
+use fuels_types::param_types::ParamType;
+use fuels_types::{TypeApplication, TypeDeclaration};
+use proc_macro2::TokenStream;
+use quote::quote;
+use std::collections::HashMap;
+use std::fmt::{Display, Formatter};
+
+// Represents a type alongside its generic parameters. Can be converted into a
+// `TokenStream` via `.into()`.
+#[derive(Debug, Clone)]
+pub(crate) struct ResolvedType {
+    pub type_name: TokenStream,
+    pub generic_params: Vec<ResolvedType>,
+    pub param_type: ParamType,
+}
+
+impl Display for ResolvedType {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", TokenStream::from(self.clone()).to_string())
+    }
+}
+
+impl From<&ResolvedType> for TokenStream {
+    fn from(resolved_type: &ResolvedType) -> Self {
+        let type_name = &resolved_type.type_name;
+        if resolved_type.generic_params.is_empty() {
+            return quote! { #type_name };
+        }
+
+        let generic_params = resolved_type
+            .generic_params
+            .iter()
+            .map(|generic_type| TokenStream::from(generic_type));
+
+        quote! { #type_name<#( #generic_params ),*> }
+    }
+}
+
+impl From<ResolvedType> for TokenStream {
+    fn from(resolved_type: ResolvedType) -> Self {
+        (&resolved_type).into()
+    }
+}
+
+/// Given a type, will recursively proceed to resolve it until it results in a
+/// `ResolvedType` which can be then be converted into a `TokenStream`. As such
+/// it can be used whenever you need the Rust type of the given
+/// `TypeApplication`.
+pub(crate) fn resolve_type(
+    type_application: &TypeApplication,
+    types: &HashMap<usize, TypeDeclaration>,
+) -> Result<ResolvedType, Error> {
+    let recursively_resolve = |type_applications: &Option<Vec<TypeApplication>>| {
+        type_applications
+            .iter()
+            .flatten()
+            .map(|array_type| resolve_type(&array_type, &types))
+            .collect::<Result<Vec<_>, _>>()
+    };
+
+    let base_type = types.get(&type_application.type_id).unwrap();
+    let param_type = ParamType::from_type_declaration(base_type, types)?;
+
+    let (type_name, generic_params) = match &param_type {
+        ParamType::Generic(name) => Ok::<_, Error>((name.parse().unwrap(), vec![])),
```

Comment:
Could make our lives easier if we're debugging something related to this:

```suggestion
        ParamType::Generic(name) => Ok::<_, Error>((name.parse().expect("Failed to parse generic param name"), vec![])),
```